### PR TITLE
WIP: remove insecure-port and deprecated insecure serving

### PR DESCRIPTION
**Pull request by mfojtik**
**What type of PR is this?**

/kind cleanup
/kind deprecation

**What this PR does / why we need it**:

This removes the `insecure-port` option and the already deprecated insecure serving for kube apiserver, kube controller manager and scheduler.
This means the health checks are possible only via secure port.

**Does this PR introduce a user-facing change?**:
YES

xref: https://github.com/kubernetes/kubernetes/issues/91506

```release-note
ACTION REQUIRED: Deprecated `insecure-port` flag has been removed from kube-apiserver, kube-controller-manager and kube-scheduler binaries. Insecure serving is no longer possible and all API servers have to use secure serving (HTTPS).
```

/cc @deads2k @sttts

### CODE REVIEW

**Comment by sttts** [comment]
Sgtm.

**Comment by mfojtik** [comment]
this will require a little plumbing to make all test servers serving securely :-)

File: cmd/kubeadm/app/phases/controlplane/manifests.go [hunk 1]
```diff
@@ -133,7 +133,6 @@ func CreateStaticPodFiles(manifestDir, patchesDir string, cfg *kubeadmapi.Cluste
 func getAPIServerCommand(cfg *kubeadmapi.ClusterConfiguration, localAPIEndpoint *kubeadmapi.APIEndpoint) []string {
 	defaultArguments := map[string]string{
 		"advertise-address":               localAPIEndpoint.AdvertiseAddress,
-		"insecure-port":                   "0",
```
Comment (neolit123) [inline cmd/kubeadm/app/phases/controlplane/manifests.go:136]: if the flags for KCM and scheduler are removed too, i guess we should remove them from the kubeadm manifests:
https://github.com/kubernetes/kubernetes/blob/c27c7a545b3ae2bf7e9a85830b9b83052a5ff4f9/cmd/kubeadm/app/phases/controlplane/manifests.go#L325

https://github.com/kubernetes/kubernetes/blob/c27c7a545b3ae2bf7e9a85830b9b83052a5ff4f9/cmd/kubeadm/app/phases/controlplane/manifests.go#L396

**Comment by neolit123** [comment]
> Deprecated `insecure-port` flag has been removed from kube-apiserver, kube-controller-manager and kube-scheduler binaries.

i'm pretty sire this will break a lot of user infra even if deprecated, so i recommend the release note to be prefixed with `ACTION REQUIRED:`.

/assign @liggitt 
PTAL, for review / approval

File: pkg/kubeapiserver/options/serving.go [hunk 2]
```diff
@@ -39,34 +37,3 @@ func NewSecureServingOptions() *genericoptions.SecureServingOptionsWithLoopback
...
-		BindPort:    8080,
-	}
-	return o.WithLoopback()
-}
-
-// DefaultAdvertiseAddress sets the field AdvertiseAddress if
-// unset. The field will be set based on the SecureServingOptions. If
-// the SecureServingOptions is not present, DefaultExternalAddress
-// will fall back to the insecure ServingOptions.
-func DefaultAdvertiseAddress(s *genericoptions.ServerRunOptions, insecure *genericoptions.DeprecatedInsecureServingOptions) error {
```
Comment (liggitt) [inline pkg/kubeapiserver/options/serving.go:57]: why was this function dropped? isn't it used for secure setup as well?

File: hack/make-rules/test-cmd.sh [hunk 3]
```diff
@@ -101,7 +99,7 @@ function run_kube_controller_manager() {
 # Exports:
 #   SUPPORTED_RESOURCES(Array of all resources supported by the apiserver).
 function create_node() {
-  kubectl create -f - -s "http://127.0.0.1:${API_PORT}" << __EOF__
+  kubectl create -f - -s "http://127.0.0.1:${SECURE_API_PORT}" << __EOF__
```
Comment (liggitt) [inline hack/make-rules/test-cmd.sh:102]: http + SECURE_API_PORT won't work (applies everywhere)

File: cmd/kubeadm/app/phases/controlplane/manifests.go (hunk 1 above)
Comment (liggitt) [reply cmd/kubeadm/app/phases/controlplane/manifests.go:136]: does kubeadm construct manifests for any older versions? we want to be absolutely positive we don't omit the flag to an old binary

**Comment by mfojtik** [comment]
@neolit123 thanks! updated

**Review by liggitt — COMMENTED** [review]
I know this is still in progress, wanted to leave a few notes early

I think we should consider leaving the insecure-port flag registered, marked deprecated (so it doesn't appear in help), inert (not wired to anything) and only accepting a value of '0' for several releases.

Well-behaved deployers are passing --insecure-port=0 everywhere right now in order to disable insecure serving. Removing the flag completely will force them to instantly update configs/manifests/deployment processes to omit that flag. If those updates accidentally omit the flag from a manifest for a previous version, that will *re-enable* insecure serving on that version.

I would much rather have an "only accept 0" flag and leave it silently registered for a few releases than risk an accident.

**Comment by neolit123** [comment]
@mfojtik please reference this tracking issue in the PR description:
https://github.com/kubernetes/kubernetes/issues/91506

**Comment by neolit123** [comment]
> Well-behaved deployers are passing --insecure-port=0 everywhere right now in order to disable insecure serving. Removing the flag completely will force them to instantly update configs/manifests/deployment processes to omit that flag. If those updates accidentally omit the flag from a manifest for a previous version, that will re-enable insecure serving on that version.

> I would much rather have an "only accept 0" flag and leave it silently registered for a few releases than risk an accident.

+1, this is the smoother transition.
(originally proposed [here](https://github.com/kubernetes/kubernetes/issues/91506#issuecomment-636112285))

**Comment by mfojtik** [comment]
> Well-behaved deployers are passing --insecure-port=0 everywhere right now in order to disable insecure serving. Removing the > flag completely will force them to instantly update configs/manifests/deployment processes to omit that flag. If those updates 
> accidentally omit the flag from a manifest for a previous version, that will re-enable insecure serving on that version.

agree, I will add the flag back, but it won't be wired to anything

File: hack/make-rules/test-cmd.sh (hunk 3 above)
Comment (mfojtik) [reply hack/make-rules/test-cmd.sh:102]: yeah, i want to sweep everything :-)

File: pkg/kubeapiserver/server/insecure_handler.go [hunk 4]
```diff
@@ -36,7 +55,7 @@ func BuildInsecureHandlerChain(apiHandler http.Handler, c *server.Config) http.H
 		handler = genericfilters.WithMaxInFlightLimit(handler, c.MaxRequestsInFlight, c.MaxMutatingRequestsInFlight, c.LongRunningFunc)
 	}
 	handler = genericapifilters.WithAudit(handler, c.AuditBackend, c.AuditPolicyChecker, c.LongRunningFunc)
-	handler = genericapifilters.WithAuthentication(handler, server.InsecureSuperuser{}, nil, nil)
```
[the remark this answers is not in the document]
Comment (mfojtik) [reply pkg/kubeapiserver/server/insecure_handler.go:39]: noticed it is unused now, will remove it

File: pkg/kubeapiserver/options/serving.go (hunk 2 above)
Comment (mfojtik) [reply pkg/kubeapiserver/options/serving.go:57]: it was used to set default advertise address for DeprecatedInsecureServingOptions which we don't have anymore?

Comment (liggitt) [reply pkg/kubeapiserver/options/serving.go:57]: ah, got that confused with https://github.com/kubernetes/kubernetes/blob/release-1.19/cmd/kube-apiserver/app/server.go#L595

File: cmd/kubeadm/app/phases/controlplane/manifests.go (hunk 1 above)
Comment (neolit123) [reply cmd/kubeadm/app/phases/controlplane/manifests.go:136]: looks like i forgot about the skew. kubeadm supports deploying X and X-1 of the control-plane.

this means that the version in cfg.KubernetesVersion has to be checked.
if this version of kubeadm is deploying a component version (cfg.KubernetesVersion) that no longer has the flags, the flags have to be removed from the manifests. for older versions they have to be preserved.

**Comment by liggitt** [comment]
Also, if secured kube-scheduler/kube-controller-manager is the only possible config those components support now, I _guess_ we should update the hard-coded componentstatus checkers to use the default secure ports 😞

```diff
        serversToValidate := map[string]*componentstatus.Server{
-               "controller-manager": {Addr: "127.0.0.1", Port: ports.InsecureKubeControllerManagerPort, Path: "/healthz"},
-               "scheduler":          {Addr: "127.0.0.1", Port: kubeschedulerconfig.DefaultInsecureSchedulerPort, Path: "/healthz"},
+               "controller-manager": {EnableHTTPS: true, TLSConfig: &tls.Config{InsecureSkipVerify: true}, Addr: "127.0.0.1", Port: ports.KubeControllerManagerPort, Path: "/healthz"},
+               "scheduler":          {EnableHTTPS: true, TLSConfig: &tls.Config{InsecureSkipVerify: true}, Addr: "127.0.0.1", Port: kubeschedulerconfig.DefaultKubeSchedulerPort, Path: "/healthz"},
        }
```

**Comment by neolit123** [comment]
> Also, if secured kube-scheduler/kube-controller-manager is the only possible config those components support now, I guess we should update the hard-coded componentstatus checkers to use the default secure ports disappointed

the release note should mention this change too if it's applied.
to me this seems tolerable with the deprecated state of the API, but it's certainly a breaking change.
yet, waiting for the API to be removed before we remove the insecure ports is also a problem.

**Comment by liggitt** [comment]
> but it's certainly a breaking change

if the backing components don't support serving insecurely any more (which is what this PR does), it seems less breaking than leaving it as is...

**Comment by neolit123** [comment]
> if the backing components don't support serving insecurely any more (which is what this PR does), it seems less breaking than leaving it as is...

yes, that is true.

File: cmd/kube-scheduler/app/server.go [hunk 5]
```diff
@@ -160,20 +160,6 @@ func Run(ctx context.Context, cc *schedulerserverconfig.CompletedConfig, sched *
...
 
-	// Start up the healthz server.
-	if cc.InsecureServing != nil {
-		separateMetrics := cc.InsecureMetricsServing != nil
-		handler := buildHandlerChain(newHealthzHandler(&cc.ComponentConfig, separateMetrics, checks...), nil, nil)
-		if err := cc.InsecureServing.Serve(handler, 0, ctx.Done()); err != nil {
-			return fmt.Errorf("failed to start healthz server: %v", err)
-		}
-	}
-	if cc.InsecureMetricsServing != nil {
```
Comment (mfojtik) [inline cmd/kube-scheduler/app/server.go:171]: @liggitt @deads2k @soltysh is this saying that scheduler metrics are only being served via insecure ? the `newMetricsHandler` is unused after this change... maybe I read this wrong.

Comment (soltysh) [reply cmd/kube-scheduler/app/server.go:171]: Spoke with @mfojtik on slack, it's properly handled few lines below in `newHealthzHandler` the only difference is for insecure it's possible to specify separate addresses for healthz and metrics, for secure both are served from a single address.

**Comment by neolit123** [comment]
this is required to merge first:
https://github.com/kubernetes/kubernetes/pull/94723

**Comment by MonzElmasry** [comment]
Hello :wave: ! Bug triage here, This PR has not been updated for a long time, so I'd like to check what's the status. The code freeze is starting 12.11.2020 (about 2 weeks from now) and while there is still plenty of time, we want to ensure that each PR has a chance to be merged on time.

As the PR is tagged for 1.20, is it still planned for this release?

**Comment by neolit123** [comment]
1.20 seems doable.

**Comment by cheftako** [comment]
> I think we should consider leaving the insecure-port flag registered, marked deprecated (so it doesn't appear in help), inert (not wired to anything) and only accepting a value of '0' for several releases.

Having the insecure-port wired to zero seems more vague than just removing it. Removing it sends a clear message its gone and you cannot rely on it anymore. If you have something like health checks or component status calls being sent to the insecure port they will stop working. Setting the value to 0 implies the same thing, but I think its better to be clear in this instance.

**Comment by liggitt** [comment]
> Having the insecure-port wired to zero seems more vague than just removing it. Removing it sends a clear message its gone and you cannot rely on it anymore. If you have something like health checks or component status calls being sent to the insecure port they will stop working. Setting the value to 0 implies the same thing, but I think its better to be clear in this instance.

The rationale was stated in https://github.com/kubernetes/kubernetes/pull/95522#pullrequestreview-507404322, and focused on preventing accidents in deployers stopping setting the flag to 0 too early:

> Well-behaved deployers are passing --insecure-port=0 everywhere right now in order to disable insecure serving. Removing the flag completely will force them to instantly update configs/manifests/deployment processes to omit that flag. If those updates accidentally omit the flag from a manifest for a previous version, that will re-enable insecure serving on that version.
> 
> I would much rather have an "only accept 0" flag and leave it silently registered for a few releases than risk an accident.

**Comment by ingvagabund** [comment]
@cheftako @liggitt @neolit123 @mfojtik are there other bits/blockers to discuss to have this merged? @mfojtik is busy, I was asked to politely take over. We have one week left before the code freeze.

**Comment by ingvagabund** [comment]
https://github.com/kubernetes/kubernetes/pull/96216 is doing the same for kube-controller-manager. Might be better to do it per component than to open one huge PR so we don't have to rebase so often and divide the work between individuals.

@ahg-g @alculquicondor for kube-scheduler. Are you aware of this effort?

**Comment by liggitt** [comment]
The insecure API server port was already removed in https://github.com/kubernetes/kubernetes/pull/95856 which was the priority for 1.20

I tend to agree that going component by component and neutering the insecure port is likely to be simpler

**Comment by alculquicondor** [comment]
+1 on a PR per component

**Comment by ingvagabund** [comment]
kube-scheduler bits: https://github.com/kubernetes/kubernetes/pull/96345

**Pull request closed without merging**